### PR TITLE
Support MIG parsing during CUDA context creation in UCX initialization

### PR DESCRIPTION
**Pull request by pentschev**
Some issues relating to MIG often arise because of lack of automated CI for that, as it requires specialized hardware that isn't currently available in gpuCI. This PR addresses errors reported recently, see https://github.com/rapidsai/dask-cuda/issues/583#issuecomment-1181505348 . Depends on #6678 .

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by wence-** [comment]
#6678 is now in, could you rebase to drop those commits?

**Review by wence- — CHANGES_REQUESTED** [review]
Mostly minor docstring fixes, but I think when getting the MIG devices we should do a bit more error checking if possible.

**Review by pentschev — COMMENTED** [review]
Thanks @wence- for the great review. I think I've addressed or responded to all your suggestions, would you mind taking another look when you have some time?

File: distributed/comm/tests/test_ucx.py [hunk 1]
```diff
@@ -324,16 +328,33 @@ async def test_simple(
 async def test_cuda_context(
     ucx_loop,
 ):
+    device_index = None
+    for i in range(device_get_count()):
+        if get_device_mig_mode(i)[0] == 0:
+            device_index = i
+            break
+        pytest.skip("No CUDA devices in non-MIG mode available.")
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/comm/tests/test_ucx.py:336]: Or perhaps, ["asking for forgiveness"](https://docs.python.org/3.5/glossary.html#term-eafp)

```python
try:
    device_index, *_ = (i for i in range(device_get_count()) if get_device_mig_mode(i)[0] == 0)
except ValueError:
    pytest.skip("No CUDA devices in non-MIG mode available.")
```

File: distributed/diagnostics/nvml.py [hunk 2]
```diff
@@ -113,19 +180,116 @@ def has_cuda_context():
...
+    """
+    init_once()
+    try:
+        device_index = int(device)
+        device_handle = pynvml.nvmlDeviceGetHandleByIndex(device_index)
+        uuid = pynvml.nvmlDeviceGetUUID(device_handle)
+    except ValueError:
+        uuid = device if isinstance(device, bytes) else bytes(device, "utf-8")
+
+        # Validate UUID, get index and UUID as seen with `nvidia-smi -L`
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/diagnostics/nvml.py:261]: Makes sense, thanks.

File: distributed/diagnostics/nvml.py [hunk 3]
```diff
@@ -113,19 +180,116 @@ def has_cuda_context():
...
+        uuid_handle = pynvml.nvmlDeviceGetHandleByUUID(uuid)
+        device_index = pynvml.nvmlDeviceGetIndex(uuid_handle)
+        uuid = pynvml.nvmlDeviceGetUUID(uuid_handle)
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/diagnostics/nvml.py:264]: As above, let's rely on nvml errors, thanks.

File: distributed/comm/tests/test_ucx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply distributed/comm/tests/test_ucx.py]: Thanks @wence- for the suggestion. I've included the last one for this test. I had also to do a few more changes to this test, details in the commit message of https://github.com/dask/distributed/pull/6720/commits/d5d5e71a3e8a6f459f0af58a98425e52b81ab96f .

[the remark this answers is not in the document]
Comment (wence-) [reply distributed/comm/tests/test_ucx.py]: Thanks!

**Comment by wence-** [comment]
Can you check the gpuci failures here? https://gpuci.gpuopenanalytics.com/job/dask/job/distributed/job/prb/job/distributed-prb/5098/

**Comment by pentschev** [comment]
> Can you check the gpuci failures here? https://gpuci.gpuopenanalytics.com/job/dask/job/distributed/job/prb/job/distributed-prb/5098/

Yeah, I'm still trying to figure those. Unfortunately the tests pass if you run them alone, but there seems to be some leakage from previous tests that cause that to fail. I'll keep trying.

**Comment by wence-** [comment]
> Yeah, I'm still trying to figure those. Unfortunately the tests pass if you run them alone, but there seems to be some leakage from previous tests that cause that to fail. I'll keep trying.

Seems like pyvml returns the PID of the handle in the host system, not the PID in the container :(.

(In a container)
```python
In [1]: import numba.cuda

In [2]: import pynvml

In [3]: numba.cuda.current_context()
Out[3]: <CUDA context c_void_p(94546817482160) of device 0>

In [4]: pynvml.nvmlInit()

In [5]: handle = pynvml.nvmlDeviceGetHandleByIndex(0)

In [6]: process, = pynvml.nvmlDeviceGetComputeRunningProcesses_v2(handle)

In [7]: process.pid
Out[7]: 32487

In [8]: import os

In [9]: os.getpid()
Out[9]: 8611
```
Outside the container:
```
$ nvidia-smi
...
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     32487      C   ...da/envs/ucx/bin/python3.9      303MiB |
+-----------------------------------------------------------------------------+
```

File: distributed/comm/ucx.py [hunk 4]
```diff
@@ -116,22 +130,22 @@ def init_once():
...
             )
 
-        cuda_visible_device = int(
+        cuda_visible_device = get_device_index_and_uuid(
             os.environ.get("CUDA_VISIBLE_DEVICES", "0").split(",")[0]
         )
         pre_existing_cuda_context = has_cuda_context()
-        if pre_existing_cuda_context is not False:
+        if pre_existing_cuda_context.has_context:
             _warn_existing_cuda_context(pre_existing_cuda_context, os.getpid())
```
Comment (wence-) [inline distributed/comm/ucx.py:138]:
```suggestion
            _warn_existing_cuda_context(pre_existing_cuda_context.device_info, os.getpid())
```

File: distributed/utils_test.py [hunk 5]
```diff
@@ -2148,6 +2148,12 @@ def ucx_loop():
     ucp.reset()
     loop.close()
 
+    # Reset also Distributed's UCX initialization, i.e., revert the effects of
+    # `distributed.comm.ucx.init_once()`.
+    import distributed.comm.ucx
+
+    distributed.comm.ucx.ucp = None
+
```
Comment (wence-) [inline distributed/utils_test.py:2156]: Aside, I think we should move this fixture to the `test_ucx.py` testfile and make it an `autouse=True` fixture. Then when writing a test we don't have to remember to require it.

Comment (wence-) [reply distributed/utils_test.py:2156]: Ah, OK, scratch that then.

File: distributed/utils_test.py [hunk 6]
```diff
@@ -2148,6 +2148,12 @@ def ucx_loop():
     ucp.reset()
     loop.close()
 
+    # Reset also Distributed's UCX initialization, i.e., revert the effects of
+    # `distributed.comm.ucx.init_once()`.
+    import distributed.comm.ucx
+
+    distributed.comm.ucx.ucp = None
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/utils_test.py:2155]: Is it not sufficient to just get the device associated with the current context and reset it? (As my changed suggestion does). TBH, I don't really know how MIG devices work with the primary contexts created by the runtime API, but primary context management only talks about device ids.

This is low priority though, and given the issues with containerisation around probing the resource management layer, we could just skip this test that way.

**Comment by pentschev** [comment]
Alright, I now marked both CUDA context tests to xfail and gpuCI is passing. Could you take one more look when you have the chance @wence- ?

**Comment by pentschev** [comment]
All 3 failing tests are the same as reported in https://github.com/dask/distributed/issues/7208 . I'm not sure whether codecov  is an artefact of the failing tests or something else, but I don't believe this PR has potential to affect it so heavily, so not really sure what is happening there.

**Comment by pentschev** [comment]
@fjetter @jrbourbeau wondering if you could help us figure if something is wrong with codecov, I don't believe this PR alone should decrease coverage, and if I try to open the [codecov report for this PR](https://app.codecov.io/gh/dask/distributed/pull/6720) and click on any file to expand it, nothing shows up. Is there some potential issue ongoing with codecov, or am I doing something potentially wrong?

**Comment by wence-** [comment]
> @fjetter @jrbourbeau wondering if you could help us figure if something is wrong with codecov, I don't believe this PR alone should decrease coverage, and if I try to open the [codecov report for this PR](https://app.codecov.io/gh/dask/distributed/pull/6720) and click on any file to expand it, nothing shows up. Is there some potential issue ongoing with codecov, or am I doing something potentially wrong?

For some reason (https://github.com/dask/distributed/actions/runs/3379719559/jobs/5611652549#step:20:36) the coverage failed to upload.

**Comment by wence-** [comment]
> I don't believe this PR alone should decrease coverage

All the mig-based code (which is more lines, I think) [e.g. https://github.com/dask/distributed/pull/6720/files#diff-457ccaa4d4a70945b095c0460229791e8142c269e9ec781b98ab6d1b1925fd8cR201-R217] will be untested because nothing runs with mig devices. Moreover, I think because GPUCI doesn't interact with codecov, the assumption will be that all the `comm/ucx.py` and `diagnostics/nvml.py` files have zero coverage. This PR has a net increase in the number of lines in those files (so the total number of uncovered lines goes up, and coverage goes down).

**Comment by wence-** [comment]
A big hammer would be to ignore those files, per https://docs.codecov.com/docs/ignoring-paths (although those patterns don't look like regexes to me)

```patch
diff --git a/codecov.yml b/codecov.yml
index 319545ae..bf003804 100644
--- a/codecov.yml
+++ b/codecov.yml
@@ -1,6 +1,12 @@
 codecov:
   require_ci_to_pass: yes
 
+ignore:
+  # These files exercise GPU functionality which is tested but doesn't
+  # interact with coverage reports, so we ignore them
+  - "distributed/comm/ucx.py"
+  - "distributed/diagnostics/nvml.py"
+  
 coverage:
   precision: 2
   round: down
```

**Comment by pentschev** [comment]
> All the mig-based code (which is more lines, I think) (e.g. https://github.com/dask/distributed/pull/6720/files#diff-457ccaa4d4a70945b095c0460229791e8142c269e9ec781b98ab6d1b1925fd8cR201-R217) will be untested because nothing runs with mig devices.

Right, but codecov complains the net change is of -0.75%.

> Moreover, I think because GPUCI doesn't interact with codecov, the assumption will be that all the comm/ucx.py and diagnostics/nvml.py files have zero coverage. This PR has a net increase in the number of lines in those files (so the total number of uncovered lines goes up, and coverage goes down).

So even if none of the lines added by this PR are actually seen by codecov, I wouldn't amount to 0.75% of the entire Distributed codebase. Finally, if you look at the whole codecov list many files that aren't at all related to this PR are considered (e.g., `pytorch.py` decreased by 97.73%). With that said, I believe the problem the failure of codecov uploading files sounds more plausible, as you mentioned in https://github.com/dask/distributed/pull/6720#issuecomment-1302301864 .

**Comment by pentschev** [comment]
@wence- I pushed part of the changes you suggested in https://github.com/dask/distributed/pull/6720#issuecomment-1302316138. NVML is actually tested by some of the CPU workflows [here](https://github.com/dask/distributed/blob/main/continuous_integration/environment-3.8.yaml#L30) and [here](https://github.com/dask/distributed/blob/aa1c6d858fc5a3e6b6a3276686f469ef25af14eb/continuous_integration/environment-3.9.yaml#L31), so we don't want to ignore it.

**Review by wence- — APPROVED** [review]

**Review by jacobtomlinson — APPROVED** [review]
Thanks for the review @wence-. This looks good to me, failing tests seem unrelated.

**Pull request merged**